rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -207,6 +212,46 @@ public void receiveHeaders(BtcBlock[] headers) throws IOException, BlockStoreExc
         }
     }
 
+    /**
+     * Receives only one header of serialized Bitcoin block headers and adds them to the internal BlockChain structure.
+     * @param header The bitcoin headers
+     */
+    public Integer receiveHeader(BtcBlock header) throws IOException, BlockStoreException {
+        Context.propagate(btcContext);
+        this.ensureBtcBlockChain();
+
+        long diffTimeStamp = bridgeConstants.getMinSecondsBetweenCallsToReceiveHeader();
+
+        long currentTimeStamp = rskExecutionBlock.getTimestamp(); //in seconds
+        Optional<Long> optionalLastTimeStamp = provider.getReceiveHeadersLastTimestamp();
+        if (optionalLastTimeStamp.isPresent() && (currentTimeStamp - optionalLastTimeStamp.get() < diffTimeStamp)) {
+            logger.warn("Receive header last TimeStamp less than {} milliseconds", diffTimeStamp);
+            return RECEIVE_HEADER_CALLED_TOO_SOON;
+        }
+
+        //Depth
+        StoredBlock previousBlock = btcBlockStore.get(header.getPrevBlockHash());
+        if (previousBlock == null) {
+            return RECEIVE_HEADER_CANT_FOUND_PREVIOUS_BLOCK;
+        }
+
+        // height of best chain - height of current header block greater than maximum depth accepted
+        if ((getBtcBlockchainBestChainHeight() - (previousBlock.getHeight() + 1)) > bridgeConstants.getMaxDepthBlockchainAccepted()) {
+            return RECEIVE_HEADER_BLOCK_TOO_OLD;
+        }
+
+        try {
+            btcBlockChain.add(header);
+        } catch (Exception e) {
+            // If we tray to add an orphan header bitcoinj throws an exception
+            // This catches that case and any other exception that may be thrown
+            logger.warn("Exception adding btc header {}", header.getHash(), e);
+            return RECEIVE_HEADER_UNEXPECTED_EXCEPTION;
+        }
+        provider.setReceiveHeadersLastTimestamp(currentTimeStamp);
```

Comment:
I feel like it would be a safe improvement to update the timestamp only when the chainhead changes and not on every valid header.

Reply:
In particular, if a block was already registered, it would update the timestamp without updating the blockchain making a DoS to this method. Checking that the block isn't know would be probably enough for this issue